employee-rostering-backend/src/main/java/org/optaweb/employeerostering/service/SkillService.java

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2019 Red Hat, Inc. and/or its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.optaweb.employeerostering.service;
+
+import java.util.List;
+import java.util.Optional;
+
+import javax.persistence.EntityExistsException;
+import javax.persistence.EntityNotFoundException;
+import javax.transaction.Transactional;
+
+import org.optaweb.employeerostering.domain.Skill;
+import org.optaweb.employeerostering.persistence.SkillRepository;
+import org.springframework.stereotype.Service;
+
+@Service
+public class SkillService extends AbstractRestService {
+
+    private final SkillRepository skillRepository;
+
+    public SkillService(SkillRepository skillRepository) {
+        this.skillRepository = skillRepository;
+    }
+
+    @Transactional
+    public List<Skill> getSkillList(Integer tenantId) {
+        return skillRepository.findAllByTenantId(tenantId);
+    }
+
+    @Transactional
+    public Skill getSkill(Integer tenantId, Long id) {
+        Optional<Skill> skillOptional = skillRepository.findById(id);
+
+        if (!skillOptional.isPresent()) {
+            throw new EntityNotFoundException("No Skill entity found with ID (" + id + ").");
+        }
+
+        validateTenantIdParameter(tenantId, skillOptional.get());
+        return skillRepository.findById(id).get();
+    }
+
+    @Transactional
+    public Boolean deleteSkill(Integer tenantId, Long id) {
+        Optional<Skill> skillOptional = skillRepository.findById(id);
+
+        if (!skillOptional.isPresent()) {
+            throw new EntityNotFoundException("No Skill entity found with ID (" + id + ").");
+        }
+
+        validateTenantIdParameter(tenantId, skillOptional.get());
+        skillRepository.deleteById(id);
+        return true;
+    }
+
+    @Transactional
+    public Skill createSkill(Integer tenantId, Skill skill) {
+        validateTenantIdParameter(tenantId, skill);
+
+        String name = skill.getName();
+
+        //If name is null
+        if (name == null) {
+            throw new IllegalStateException("Skill entity name with tenantId (" + tenantId + ") cannot be null.");
+        }
+
+        List<Skill> skillList = skillRepository.findAllByTenantId(tenantId);
+
+        for (Skill s : skillList) {
+            //If skill already exists
+            if (s.getName().equals(name)) {
+                throw new EntityExistsException("Skill entity with tenantId (" + tenantId + ") and name (" + name
```

Comment:
Is there no way to set constraints directly on entities? In the old backend, we had
```
@Table(uniqueConstraints = @UniqueConstraint(columnNames = {"tenantId", "name"}))
public class Skill extends AbstractPersistable {
// ...
```
which creates a unique constraint for the pair `(tenantId, name)`, and will throw an exception automatically if you try to persist/update name to an existing one for the tenant.

We also have an `@AssertTrue` for different kind of constraints (ex: start time before end time in `EmployeeAvaliability`)

Reply:
@Christopher-Chianelli +1. This applies for all kinds of validation (nonNull ...). Check https://www.baeldung.com/javax-validation.

Reply:
You're right, @UniqueConstraint is a better way to handle duplicates.